File: common/a2a/protocol.py
from typing import Union, Any, List, Literal, Optional
from pydantic import BaseModel, Field, field_serializer
from datetime import datetime
from uuid import uuid4
from enum import Enum


class TaskState(str, Enum):
    SUBMITTED = "submitted"
    WORKING = "working"
    INPUT_REQUIRED = "input-required"
    COMPLETED = "completed"
    CANCELED = "canceled"
    FAILED = "failed"
    UNKNOWN = "unknown"


class TextPart(BaseModel):
    type: Literal["text"] = "text"
    text: str
    metadata: dict[str, Any] | None = None


class DataPart(BaseModel):
    type: Literal["data"] = "data"
    data: dict[str, Any]
    metadata: dict[str, Any] | None = None


Part = Union[TextPart, DataPart]


class Message(BaseModel):
    role: Literal["user", "agent"]
    parts: List[Part]
    metadata: dict[str, Any] | None = None


class TaskStatus(BaseModel):
    state: TaskState
    message: Message | None = None
    timestamp: datetime = Field(default_factory=datetime.now)
    
    @field_serializer('timestamp')
    def serialize_timestamp(self, dt: datetime, _info):
        """Serialize datetime to ISO format string"""
        return dt.isoformat()


class Artifact(BaseModel):
    name: str | None = None
    description: str | None = None
    parts: List[Part]
    metadata: dict[str, Any] | None = None


class Task(BaseModel):
    id: str
    sessionId: str | None = None
    status: TaskStatus
    artifacts: List[Artifact] | None = None
    history: List[Message] | None = None
    metadata: dict[str, Any] | None = None


class TaskStatusUpdateEvent(BaseModel):
    id: str
    status: TaskStatus
    final: bool = False
    metadata: dict[str, Any] | None = None


class TaskArtifactUpdateEvent(BaseModel):
    id: str
    artifact: Artifact    
    metadata: dict[str, Any] | None = None


class TaskIdParams(BaseModel):
    id: str
    metadata: dict[str, Any] | None = None


class TaskQueryParams(TaskIdParams):
    historyLength: int | None = None


class TaskSendParams(BaseModel):
    id: str = Field(default_factory=lambda: uuid4().hex)
    sessionId: str = Field(default_factory=lambda: uuid4().hex)
    message: Message
    historyLength: int | None = None
    metadata: dict[str, Any] | None = None


# JSON-RPC classes
class JSONRPCMessage(BaseModel):
    jsonrpc: Literal["2.0"] = "2.0"
    id: str = Field(default_factory=lambda: uuid4().hex)


class JSONRPCRequest(JSONRPCMessage):
    method: str
    params: dict[str, Any] | None = None


class JSONRPCError(BaseModel):
    code: int
    message: str
    data: Any | None = None


class JSONRPCResponse(JSONRPCMessage):
    result: Any | None = None
    error: JSONRPCError | None = None


class SendTaskRequest(JSONRPCRequest):
    method: Literal["tasks/send"] = "tasks/send"
    params: TaskSendParams


class SendTaskResponse(JSONRPCResponse):
    result: Task | None = None


# Streaming classes
class SendTaskStreamingRequest(JSONRPCRequest):
    method: Literal["tasks/send/stream"] = "tasks/send/stream"
    params: TaskSendParams


class SendTaskStreamingResponse(JSONRPCResponse):
    result: TaskStatusUpdateEvent | TaskArtifactUpdateEvent | None = None


# Error types
class JSONParseError(JSONRPCError):
    code: int = -32700
    message: str = "Invalid JSON payload"
    data: Any | None = None


class InvalidRequestError(JSONRPCError):
    code: int = -32600
    message: str = "Request payload validation error"
    data: Any | None = None


class MethodNotFoundError(JSONRPCError):
    code: int = -32601
    message: str = "Method not found"
    data: None = None


class InvalidParamsError(JSONRPCError):
    code: int = -32602
    message: str = "Invalid parameters"
    data: Any | None = None


class InternalError(JSONRPCError):
    code: int = -32603
    message: str = "Internal error"
    data: Any | None = None


class AgentSkill(BaseModel):
    id: str
    name: str
    description: str | None = None
    examples: List[str] | None = None


class AgentCard(BaseModel):
    name: str
    description: str | None = None
    url: str
    version: str
    skills: List[AgentSkill]


class A2AClientError(Exception):
    pass


class A2AClientHTTPError(A2AClientError):
    def __init__(self, status_code: int, message: str):
        self.status_code = status_code
        self.message = message
        super().__init__(f"HTTP Error {status_code}: {message}")


class A2AClientJSONError(A2AClientError):
    def __init__(self, message: str):
        self.message = message
        super().__init__(f"JSON Error: {message}") 

File: common/js/protocol.ts
/**
 * A2A Protocol TypeScript Implementation
 * Based on the Python implementation and A2A specification
 */

export enum TaskState {
  SUBMITTED = "submitted",
  WORKING = "working",
  INPUT_REQUIRED = "input-required",
  COMPLETED = "completed",
  CANCELED = "canceled",
  FAILED = "failed",
  UNKNOWN = "unknown"
}

export type TextPart = {
  type: "text";
  text: string;
  metadata?: Record<string, unknown> | null;
};

export type DataPart = {
  type: "data";
  data: Record<string, unknown>;
  metadata?: Record<string, unknown> | null;
};

export type FilePart = {
  type: "file";
  file: {
    bytes?: string | null;
    uri?: string | null;
  };
  metadata?: Record<string, unknown> | null;
};

export type Part = TextPart | DataPart | FilePart;

export type Message = {
  role: "user" | "agent";
  parts: Part[];
  metadata?: Record<string, unknown> | null;
};

export type TaskStatus = {
  state: TaskState;
  message?: Message | null;
  timestamp: string;
};

export type Artifact = {
  name?: string | null;
  description?: string | null;
  parts: Part[];
  metadata?: Record<string, unknown> | null;
};

export type Task = {
  id: string;
  sessionId?: string | null;
  status: TaskStatus;
  artifacts?: Artifact[] | null;
  history?: Message[] | null;
  metadata?: Record<string, unknown> | null;
};

export type TaskStatusUpdateEvent = {
  id: string;
  status: TaskStatus;
  final: boolean;
  metadata?: Record<string, unknown> | null;
};

export type TaskArtifactUpdateEvent = {
  id: string;
  artifact: Artifact;
  metadata?: Record<string, unknown> | null;
};

export type TaskIdParams = {
  id: string;
  metadata?: Record<string, unknown> | null;
};

export type TaskQueryParams = TaskIdParams & {
  historyLength?: number | null;
};

export type TaskSendParams = {
  id?: string;
  sessionId?: string;
  message: Message;
  historyLength?: number | null;
  metadata?: Record<string, unknown> | null;
};

// JSON-RPC classes
export type JSONRPCMessage = {
  jsonrpc: "2.0";
  id?: string | number | null;
};

export type JSONRPCRequest = JSONRPCMessage & {
  method: string;
  params?: Record<string, unknown> | null;
};

export type JSONRPCError = {
  code: number;
  message: string;
  data?: unknown | null;
};

export type JSONRPCResponse = JSONRPCMessage & {
  result?: unknown | null;
  error?: JSONRPCError | null;
};

export type SendTaskRequest = JSONRPCRequest & {
  method: "tasks/send";
  params: TaskSendParams;
};

export type SendTaskResponse = JSONRPCResponse & {
  result?: Task | null;
};

// Streaming classes
export type SendTaskStreamingRequest = JSONRPCRequest & {
  method: "tasks/send/stream";
  params: TaskSendParams;
};

export type SendTaskStreamingResponse = JSONRPCResponse & {
  result?: TaskStatusUpdateEvent | TaskArtifactUpdateEvent | null;
};

// Get task request/response
export type GetTaskRequest = JSONRPCRequest & {
  method: "tasks/get";
  params: TaskQueryParams;
};

export type GetTaskResponse = JSONRPCResponse & {
  result?: Task | null;
};

// Error types
export const JSONRPCErrorCodes = {
  PARSE_ERROR: -32700,
  INVALID_REQUEST: -32600,
  METHOD_NOT_FOUND: -32601,
  INVALID_PARAMS: -32602,
  INTERNAL_ERROR: -32603
};

export type AgentSkill = {
  id: string;
  name: string;
  description?: string | null;
  examples?: string[] | null;
};

export type AgentCard = {
  name: string;
  description?: string | null;
  url: string;
  version: string;
  skills: AgentSkill[];
};

// Error classes
export class A2AClientError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "A2AClientError";
  }
}

export class A2AClientHTTPError extends A2AClientError {
  statusCode: number;
  
  constructor(statusCode: number, message: string) {
    super(`HTTP Error ${statusCode}: ${message}`);
    this.name = "A2AClientHTTPError";
    this.statusCode = statusCode;
  }
}

export class A2AClientJSONError extends A2AClientError {
  constructor(message: string) {
    super(`JSON Error: ${message}`);
    this.name = "A2AClientJSONError";
  }
} 